### PR TITLE
(#52) Allow user effect header manually

### PR DESCRIPTION
**Pull request by shana0440**

### CODE REVIEW

File: response.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	headers []Header
+	body    interface{}
+}
+
+type Header struct {
+	Key   string
+	Value string
+}
+
+func NewResponse(body interface{}) *Response {
```
Comment (dannypsnl) [inline response.go]: Ok, I understand why `interface{}` now. We don't have to handle it now.
But `Content-Type` is a real problem here. We can receive a JSON but response a pure text.

File: server_test.go [hunk 2]
```diff
@@ -60,13 +60,20 @@ var (
 		}
 		return "cookies"
 	})
+	customResponseHeader = rocket.Get("/", func() *rocket.Response {
+		body := rocket.Json(`{"msg": "welcome"}`)
+		return rocket.NewResponse(body).Headers(
```
[the remark this answers is not in the document]
Comment (dannypsnl) [reply server_test.go]: Leave a comment at `NewResponse` there.

[the remark this answers is not in the document]
Comment (shana0440) [reply server_test.go]: What your mean?

File: response.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package rocket
+
+type Response struct {
+	headers []Header
```
Comment (dannypsnl) [inline response.go]: Headers -> `map[string]string`, for good to use & for performance.

File: response.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+func NewResponse(body interface{}) *Response {
+	return &Response{
+		headers: make([]Header, 0),
+		body:    body,
+	}
+}
+
+func (res *Response) Headers(headers ...Header) *Response {
```
Comment (dannypsnl) [inline response.go]: Then you don't have to append at here.

File: rocket.go [hunk 5]
```diff
@@ -63,9 +63,18 @@ func (rk *Rocket) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	}
-	w.Header().Set("Content-Type", contentTypeOf(response))
-
-	fmt.Fprint(w, response)
+	switch response.(type) {
+	case *Response:
+		res := response.(*Response)
+		w.Header().Set("Content-Type", contentTypeOf(res.body))
+		for _, header := range res.headers {
+			w.Header().Set(header.Key, header.Value)
```
Comment (dannypsnl) [inline rocket.go]: And you can `for k, v := range res.headers` here.

File: response.go (hunk 4 above)
Comment (shana0440) [reply response.go]: resolve at 6431581b27589646600d86deb8783690ff15ddf3

File: response.go [hunk 6]
```diff
@@ -12,12 +12,14 @@ type Header struct {
 
 func NewResponse(body interface{}) *Response {
 	return &Response{
-		headers: make([]Header, 0),
+		headers: make(map[string]string),
 		body:    body,
 	}
 }
 
 func (res *Response) Headers(headers ...Header) *Response {
```
Comment (dannypsnl) [inline response.go]: I would say have a type `Headers` at here would be better. Just use it like:
```go
rocket.Headers{
    "Content-Type": "application/json",
    "Allow-Origin": "*",
}
```

File: server_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (dannypsnl) [reply server_test.go]: @shana0440 https://github.com/dannypsnl/rocket/pull/66#discussion_r217951154

File: response.go (hunk 1 above)
Comment (dannypsnl) [reply response.go]: I sought that content-type, maybe ok I think.

Comment (dannypsnl) [reply response.go]: LGTM

File: response.go (hunk 6 above)
Comment (shana0440) [reply response.go]: Yes, It looks better, but I need to think the other name, `rocket.Headers` already used.

Comment (dannypsnl) [reply response.go]: I think use `Header` is ok, or create a sub pkg: `header`

Comment (dannypsnl) [reply response.go]: And I have a question, when would we need to read the header manually? Except content-type?

Comment (shana0440) [reply response.go]: like CSRF token, Authorization or any custom header.

**Comment by dannypsnl** [comment]
Then I believe we need to add godoc for Header & Headers, else they are pretty confused people.

**Pull request merged**